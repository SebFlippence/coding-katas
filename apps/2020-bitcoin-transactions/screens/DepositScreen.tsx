import * as React from 'react';
import { StyleSheet, Text, TextInput, View, Button } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { api } from '../constants/config';

export default function DepositScreen() {
  const [value, onChangeText] = React.useState('');
  const { navigate } = useNavigation();

  const onSubmit = async () => {
    if (value === '') return;

    await fetch(`${api.hostname}${api.transactions}`, {
      method: 'POST',
      mode: 'no-cors',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        currency: 'BÃ¸',
        amount: value,
        counterparty: 'Bit Credit',
        category: 'Gifts',
      }),
    });

    onChangeText('');

    navigate('Transactions');
  };

  return (
    <View style={styles.container}>
      <Text style={[styles.row, styles.title]}>Add Bitcoin</Text>
      <TextInput
        onChangeText={text => onChangeText(text)}
        onFocus={() => onChangeText('')}
        value={value}
        keyboardType="number-pad"
        style={[styles.row, styles.input]}
        onSubmitEditing={onSubmit}
      />
      <Button title="Add" onPress={onSubmit} />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fdfdfd',
    padding: 15,
    flexDirection: 'column',
    alignItems: 'center',
  },
  row: {
    margin: 10,
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
  },
  input: { height: 40, borderColor: 'gray', borderWidth: 1, width: 100 },
});
