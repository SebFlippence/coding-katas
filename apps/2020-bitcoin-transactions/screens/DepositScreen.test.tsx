import * as React from 'react';
import renderer from 'react-test-renderer';

import DepositScreen from './DepositScreen';

const { act } = renderer;

jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn().mockImplementation(() => ({
    navigate: jest.fn(),
  })),
}));

describe('DepositScreen', () => {
  beforeEach(() => {
      global.fetch = jest.fn().mockImplementation(
        () =>
          new Promise(resolve =>
            resolve({
              ok: true,
              json() {
                return { itineraries: [], legs: [] };
              },
            })
          )
      );
  });

  test('rendering', () => {
    const tree = renderer.create(<DepositScreen />);
    expect(tree).toMatchSnapshot();
  });

  test('adding credit', async () => {
    const wrapper = renderer.create(<DepositScreen />);

    act(() => {
      const { onChangeText } = wrapper.root.findByType('TextInput').props;
      onChangeText('1.23');
    });

    await act(async () => {
      const { onSubmitEditing } = wrapper.root.findByType('TextInput').props;
      await onSubmitEditing();
    });

    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith('http://localhost:8080/transactions', {
      body: '{"currency":"BÃ¸","amount":"1.23","counterparty":"Bit Credit","category":"Gifts"}',
      headers: { 'Content-Type': 'application/json' },
      method: 'POST',
      mode: 'no-cors',
    });
  });
});
