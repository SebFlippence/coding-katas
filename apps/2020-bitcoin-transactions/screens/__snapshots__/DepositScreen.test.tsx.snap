// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DepositScreen rendering 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fdfdfd",
      "flex": 1,
      "flexDirection": "column",
      "padding": 15,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "margin": 10,
        },
        Object {
          "fontSize": 20,
          "fontWeight": "bold",
        },
      ]
    }
  >
    Add Bitcoin
  </Text>
  <TextInput
    allowFontScaling={true}
    keyboardType="number-pad"
    onChangeText={[Function]}
    onFocus={[Function]}
    onSubmitEditing={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "margin": 10,
        },
        Object {
          "borderColor": "gray",
          "borderWidth": 1,
          "height": 40,
          "width": 100,
        },
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Add
      </Text>
    </View>
  </View>
</View>
`;
